test_runner/regress/test_pageserver_generations.py

```diff
@@ -668,3 +669,22 @@ def test_upgrade_generationless_local_file_paths(
         )
         == 0
     )
+
+    # Check that when we evict and promote one of the legacy-named layers, everything works as
+    # expected
+    local_layers = list(
+        (
+            parse_layer_file_name(path.name),
+            os.path.join(env.pageserver.timeline_dir(tenant_id, timeline_id), path),
+        )
+        for path in env.pageserver.list_layers(tenant_id, timeline_id)
+    )
+    (victim_layer_name, victim_path) = local_layers[0]
+    assert os.path.exists(victim_path)
+
+    env.pageserver.http_client().evict_layer(tenant_id, timeline_id, victim_layer_name.to_str())
+    assert not os.path.exists(victim_path)
+
+    env.pageserver.http_client().download_layer(tenant_id, timeline_id, victim_layer_name.to_str())
+    # We should download into the same local path we started with
+    assert os.path.exists(victim_path)
```

Comment:
This test will have to updated after un-comment the code in `local_layer_path`, right?

Reply:
We already uncommented that in #7660 

The download goes into the same path because the `Layer` object stores the path it was constructed with, so even when we evict/download, it keeps using the legacy path (which is functionally fine, but a bit annoying).  

This line _would_ need updating if we refactor the layer stuff to only store the local path when we actually have a downloaded layer, but to do so is a bit involved because there are currently some code paths that expect to get the layer path whether it's downloaded or not.

Reply:
Ugh. I was looking at an old checkout. I see that makes sense. Let me check the startup reconciliation logic again